### PR TITLE
Prefer factories over prototype + this

### PR DESCRIPTION
**Pull request by thepatrick**
We don't use the prototype much, and generally prefer the pattern of calling a method and being given an object back. Let's formalise that.

## What

```javascript
// before
function Widget(name) {
  this.privateField = 'Hello ';
  this.name = name;
}
var foo = new Thing();

// after
var createWidget = function(name) {
  var api = new (function Widget(){})();

  var privateField = 'Hello ';
  api.name = name;

  return api;
}
var foo = createThing();
```

## Why

1. Using thew `new` keyword and an object prototype makes it harder to separate internal properties and functions from intentionally external APIs
2. We typically do this anyway
3. We avoid using `this` in new code anyway

## Up for debate

1. `var api...` The name of this variable. In #1 @aoberoi argued for it being `self`. I'd like us to make a decision on that name (keeping in mind that it will apply for this one too)

2. `var api = new (function Widget(){})();` This may help us in debuggers, but is it worth the overhead? Should this be noted as an optional suggestion? Only for publicly returned objects?

### CODE REVIEW

**Comment by ggarber** [comment]
Stupid comment.  What's the problem with this:

```
function Widget(name) {
  var privateField = 'Hello ';
  this.name = name;
}
var foo = new Thing();
```

**Comment by thepatrick** [comment]
@ggarber `this` is the problem: it's too easy to forget to `bind` a function (which creates a second function) or to assign the `this` to something else (like `var _this = this;`), and then it becomes hard to keep track of.

To show why lets use a more contrived example:

```javascript
function Widget(name) {
  var privateField = 'Hello ';
  this.name = name;
  this.speakWhenClicked = function(element) {
    element.addEventListener('click', function() {
      alert('Hello ' + this.name);
    });  
  };
}
var foo = new Thing();
```
Which of course will alert 'Hello undefined'. In the alternative world:

```javascript
var createWidget = function(name) {
  var api = {};
  var privateField = 'Hello ';
  api.name = name;
  this.speakWhenClicked = function(element) {
    element.addEventListener('click', function() {
      alert('Hello ' + api.name);
    });  
  };
};
```

that can't happen. At least not for the same reason :)

**Comment by jraoult** [comment]
I love the direction this thing is taking.

Some ideas / questions:
* What about using `Object.freeze`? Specially for public facing API.
* Also do we agree that we don't have to use the name `api` but we can use a better tailored name that matches what we are creating?
* Finally, why `new (function Widget(){})();`? Why not a simple object literal?

All that would give something like:
```javascript
function createWidget() {

  function soCalledPrivateFunction() {
    aProperty++;
  }

  function toBeExposed() {
    soCalledPrivateFunction();
  }

  var aProperty = 0;

  var widget = {
    toBeExposed: toBeExposed
    get aProperty() {
      return aProperty;
    }
  };

  return Object.freeze(widget);
}
```

**Comment by ggarber** [comment]
@thepatrick  Thank you for the explanation.   I'm not convinced but it is ok :)

Anyways it is funny that in your example of how easy is to make mistakes with option 1 you made two mistakes with option 2 :P (i think).

**Comment by luma** [comment]
@jraoult 

| What about using Object.freeze? Specially for public facing API.
@thepatrick and I talked about that. I think it's a good idea

| Also do we agree that we don't have to use the name api but we can use a better tailored name that matches what we are creating?
I don't agree ;-) For me the better tailored name comes from the named function that api instantiates. That's what you'll see in the debugger. We should use a generic name as what it models is generic: it's always the public API of whatever named object this is.

I don't think it needs to be called api, I do think it should always be called the same thing.

| Finally, why new (function Widget(){})();? Why not a simple object literal?
For the reasons above. It's the actual identity of the "type" of object. It also makes debugging easier.

**Comment by luma** [comment]
> @thepatrick Thank you for the explanation. I'm not convinced but it is ok :)
>
> Anyways it is funny that in your example of how easy is to make mistakes with option 1 you made two mistakes with option 2 :P (i think).

@thepatrick @ggarber: The problem is that these examples are kept simple so that they are easy to consume. Where it really pays off is in larger objects. For example, I converted all the PeerConnection and Media code to use this pattern in some experimental branches and found quite dramatic increases in readability and decreases in certain types of coding errors.

This was mostly because this pattern eliminates needing to reason about or manipulate this. Somethat that becomes complex as the object increase in size/complexity. For examples of this complexity look at session.js, publisher.js, subscriber.js, etc

**Comment by thepatrick** [comment]
@ggarber fixed the minor issues, that actually would have been caught by jshint/jscs.

**Comment by jraoult** [comment]
>  don't agree ;-) For me the better tailored name comes from the named function that api instantiates. That's what you'll see in the debugger. We should use a generic name as what it models is generic: it's always the public API of whatever named object this is.

Even if the public name of the thing is defined some where else it shouldn't stop us to use a "real" name internally. To me it is the same reasoning as for #1 where I think a meaningful name even if "local" only brings more context to the reader than `self` or `_this`. Let's agree to disagree on this one, will see what the others think.

> For the reasons above. It's the actual identity of the "type" of object. It also makes debugging easier.

I see what you mean. It is an ugly boilerplate though, let's see if we can find something else.

**Comment by jraoult** [comment]
@thepatrick regarding your RFC, I told what I thought but at the end as long as we agree on something and this something is better than the current status quo, I'm fine. I would be happy to know what @leahciMic @voltrevo @aullman think.

**Comment by luma** [comment]
+1 in general. This is the pattern that I already used. I like just using `api` for the reasons already stated but I am not wedded to that if everyone else thinks it's a bad idea

**Comment by leahciMic** [comment]
This is pretty much the revealing module pattern. I have no problems with it. However, are we advocating this is the only pattern we can use in our projects?

A good read for common design patterns for JavaScript: http://www.addyosmani.com/resources/essentialjsdesignpatterns/book/#designpatternsjavascript

**Comment by jraoult** [comment]
@leahciMic I think so, the goal is to have a common way to do things so that we don't have to understand how a particular piece of code works regarding instantiation and we can spend more "brain cycles" on the "business logic".

That said, if your concern is that this pattern prevents us to do some things it is important to raise it now so that we can refine it.

**Comment by luma** [comment]
@leahciMic No, more like saying that this is a sensible default for a great many situations. If you've got a good reason for doing it another way then great, but the reason needs to be something other than taste ;-)

TLDR: We are standardising on common usage. There will always be edge cases where more specific patterns should be used, and that's OK.

**Comment by thepatrick** [comment]
@leahciMic you correct in that it's the revealing module pattern.

Deviation from the style guide is an option - but it'd need to be justified.
